source/common/network/connection_impl.cc

```diff
@@ -824,6 +824,8 @@ ClientConnectionImpl::ClientConnectionImpl(
     : ConnectionImpl(dispatcher, std::make_unique<ClientSocketImpl>(remote_address, options),
                      std::move(transport_socket), stream_info_, false),
       stream_info_(dispatcher.timeSource(), socket_->addressProviderSharedPtr()) {
+
+  stream_info_.setUpstreamConnectionId(id());
```

Comment:
I'm not sure you need this.

Since we have two streamInfo for two sides. One for downstream, another one for upstream. We copy the upstream streamInfo 's id and SSL connection info into the downstream's streamInfo.  This is done into UpstreamRequest

So here, you set the upstream's connection id into the upstream's streamInfo, it seems needless.

Reply:
@soulxu I am not sure I understand your comment. The value that gets assigned in onPoolReady in UpstreamRequest comes from the value set here. So we definitely need to set the value in the ConnectionImpl instance variable.

However, I do hear your comment about the confusion with upstream/downstream in the instance of stream info. In other places in the codebase stream info does have a defined upstream/downstream directionality. Given I am implementing this in the ClientConnection class I thought upstream was appropriate. I am going to add some more clarifying comment per @alyssawilk's [comment](https://github.com/envoyproxy/envoy/pull/17512#discussion_r681901853) above

Reply:
> @soulxu I am not sure I understand your comment. The value that gets assigned in onPoolReady in UpstreamRequest comes from the value set here. So we definitely need to set the value in the ConnectionImpl instance variable.
> 

Sorry, I just found you set the upstreamId of upstream's StreamInfo to the downstreamId of downstream's StreamInfo. That is opposite with the SSL connection info which above your change:

https://github.com/envoyproxy/envoy/blob/a8a2eb9fecb45e0e3e8a69f8176d801faed33c1c/source/common/router/upstream_request.cc#L420-L421

you can see the upstream SSL connection info of downstream's streamInfo is coming from the downstream SSL connection of upstream's streamInfo.

Yes, as @alyssawilk said, we interleave the upstream and downstream here. But this will be more confusing, since we have a different pattern between SSL connection info and connectionId.

Reply:
Actually, the connectionID already set at https://github.com/envoyproxy/envoy/blob/b29d6543e7568a8a3e772c7909a1daa182acc670/source/common/network/connection_impl.cc#L101

So what you need to do in the onPoolReady in upstreamRequest is:
`stream_info_.setUpstreamConnectionId(info.downstreamAddressProvider().connectionID());`

Reply:
actually this looks like it'd be a better option.  sorry to not merge this but I think it's worth the fix.
cc @goaway in case someone on the mobile side was waiting on this while Jose is out.
/wait

Reply:
> we interleave the upstream and downstream here. But this will be more confusing, since we have a different pattern between SSL connection info and connectionId.

@soulxu yep, I agree. I chose to use upstream here because of two reasons:
1. It made more sense to me, as really the connection is to an upstream entity.
2. I did not want to add downstreamConnectionId as well.

However, I agree with you that that creates confusion, and thanks to your suggestion we can just skip that :). Thanks a lot for taking the time to review and point to a better solution. I appreciate it!

I have updated to use the address provider.